### PR TITLE
feat: add base scheduler application

### PR DESCRIPTION
**Pull request by iverly**
This PR adds the basic application for the scheduler.

Explanations:

The main function creates an instance of the `Manager` structure.
Who is responsible for managing startup, data, and channels between all threads.

Currently, the manager only creates the gRPC server and initializes the storage.

gRPC services return a default response for all functions.

### CODE REVIEW

File: scheduler/src/manager.rs [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+            nodes: Storage::new(),
+        }
+    }
+
+    /// This function returns a reference to the instances storage.
+    ///
+    /// Returns:
+    ///
+    /// A reference to the instances storage.
+    pub fn get_instances_storage(&self) -> &Storage<Instance> {
```
Comment (sameo) [inline scheduler/src/manager.rs:37]: Rust getters should simply be called instances()

File: scheduler/src/manager.rs [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        &self.instances
+    }
+
+    ///This function returns a reference to the nodes storage.
+    ///
+    /// Returns:
+    ///
+    /// A reference to the nodes storage.
+    pub fn get_nodes_storage(&self) -> &Storage<Node> {
```
Comment (sameo) [inline scheduler/src/manager.rs:46]: Ditto: nodes()

File: scheduler/src/manager.rs [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        &self.nodes
+    }
+
+    /// > This function creates a gRPC server that listens on port 50051 and registers the
+    /// `NodeServiceServer` with the server
+    ///
+    /// Returns:
+    ///
+    /// A JoinHandle<()>
+    pub fn create_grpc_server(&self) -> JoinHandle<()> {
```
Comment (sameo) [inline scheduler/src/manager.rs:56]: Does this one need to be public?

File: scheduler/src/instance_listener.rs [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+#[tonic::async_trait]
+impl InstanceService for InstanceListener {
+    async fn create(
+        &self,
+        request: Request<Instance>,
+    ) -> Result<Response<Self::CreateStream>, Status> {
+        debug!("{:?}", request);
+
+        let (tx, rx) = mpsc::channel(4);
```
Comment (sameo) [inline scheduler/src/instance_listener.rs:21]: How is this 4 defined?

Comment (sameo) [inline scheduler/src/instance_listener.rs:21]: I suppose the tx half will be used in upcoming PRs?

File: scheduler/src/manager.rs (hunk 1 above)
Comment (iverly) [reply scheduler/src/manager.rs]: Fixed

File: scheduler/src/manager.rs (hunk 3 above)
Comment (iverly) [reply scheduler/src/manager.rs]: No, fixed.

File: scheduler/src/instance_listener.rs (hunk 4 above)
Comment (iverly) [reply scheduler/src/instance_listener.rs]: Yes, I've pushed more commits for you to understand our structure.

File: scheduler/src/manager.rs (hunk 2 above)
Comment (iverly) [reply scheduler/src/manager.rs]: Fixed

File: scheduler/src/instance_listener.rs (hunk 4 above)
Comment (iverly) [reply scheduler/src/instance_listener.rs]: We use the Manager::create_mpsc_channel function now here. The buffer is set to 32.

File: scheduler/src/manager.rs [hunk 5]
```diff
@@ -1,31 +1,37 @@
...
 pub struct Manager {
-    instances: Storage<Instance>,
-    nodes: Storage<Node>,
+    instances: Arc<Storage<Instance>>,
+    nodes: Arc<Storage<Node>>,
 }
 
 impl Manager {
-    /// `new` creates a new `Manager` struct with two empty `Storage` structs
+    /// The `new` function creates a new `Manager` struct with two empty `Storage` structs and a new `Orchestrator` struct.
```
[the remark this answers is not in the document]
Comment (iverly) [reply scheduler/src/manager.rs:23]: Done

File: scheduler/src/manager.rs [hunk 6]
```diff
@@ -41,13 +42,38 @@ impl Manager {
         &self.nodes
     }
 
+    /// > This function creates a gRPC server that listens on port 50051 and registers the
```
[the remark this answers is not in the document]
Comment (iverly) [reply scheduler/src/manager.rs]: Done

File: scheduler/src/instance_listener.rs [hunk 7]
```diff
@@ -7,38 +7,51 @@ use proto::scheduler::{
...
     async fn create(
         &self,
         request: Request<Instance>,
     ) -> Result<Response<Self::CreateStream>, Status> {
-        debug!("{:?}", request);
-
-        let (tx, rx) = mpsc::channel(4);
+        debug!("received request: {:?}", request);
+        let (tx, rx) = Manager::create_mpsc_channel();
+        self.sender.send(Event::InstanceCreate(request.into_inner(), tx)).await.unwrap();
```
[the remark this answers is not in the document]
Comment (iverly) [reply scheduler/src/instance_listener.rs]: Done

File: scheduler/src/manager.rs [hunk 8]
```diff
@@ -78,16 +88,87 @@ impl Manager {
...
+    /// Returns:
+    /// 
+    /// A JoinHandle<()>
+    fn listen_events(&self, mut rx: mpsc::Receiver<Event>) -> JoinHandle<()> {
+        info!("listening for incoming events ...");
+
+        tokio::spawn(async move {
+            while let Some(event) = rx.recv().await {
+                debug!("received event : {:?}", event);
+                match event {
```
[the remark this answers is not in the document]
Comment (iverly) [reply scheduler/src/manager.rs:123]: Thanks for the feedback, I've split the commits following your instructions.

**Review by sameo — APPROVED** [review]

**Pull request merged**